packages/flutter/lib/src/material/floating_action_button.dart

```diff
@@ -227,65 +227,95 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
+
+    assert(_notchAssertions(host, guest, start, end, fabRadius, notchRadius));
+
+    // If there's no overlap between the notch circle and the host, don't
+    // make a notch, just return a straight line from start to end.
+    if (!host.overlaps(guest.inflate(widget.notchMargin)))
+      return new Path()..lineTo(end.dx, end.dy);
+
+    // We build a path for the notch from 3 segments:
+    // Segment A - a Bezier curve from the host's top edge to segment B.
+    // Segment B - an arc with radius notchRadius.
+    // Segment C - a Bezier curver from segment B back to the host's top edge.
     //
-    // The other side (b) would be the distance on the horizontal axis between the
-    // notch's center and the intersection points with it's top edge.
-    final double a = host.top - guest.center.dy;
-    final double b = math.sqrt(notchRadius * notchRadius - a * a);
+    // A detailed explanation and the derivation of the formulas below is
+    // available at: https://goo.gl/Ufzrqn
+
+    const double s1 = 15.0;
+    const double s2 = 1.0;
+
+    final double r = notchRadius;
+    final double a = -1.0 * r - s2;
+    final double b = host.top - guest.center.dy;
+
+    final double n2 = math.sqrt(b * b * r * r * (a * a + b * b - r * r));
+    final double p2xA = ((a * r * r) - n2) / (a * a + b * b);
+    final double p2xB = ((a * r * r) + n2) / (a * a + b * b);
+    final double p2yA = math.sqrt(r * r - p2xA * p2xA);
+    final double p2yB = math.sqrt(r * r - p2xB * p2xB);
+
+    final List<Offset> p = new List<Offset>(6);
+
+    // p0, p1, and p2 are the control points for segment A.
+    p[0] = new Offset(a - s1, b);
+    p[1] = new Offset(a, b);
+    final double cmp = b < 0 ? -1.0 : 1.0;
+    p[2] = cmp * p2yA > cmp * p2yB ? new Offset(p2xA, p2yA) : new Offset(p2xB, p2yB);
+
+    // p3, p4, and p5 are the control points for segment B, which is a mirror
+    // of segment A around the y axis.
+    p[3] = new Offset(-1.0 * p[2].dx, p[2].dy);
+    p[4] = new Offset(-1.0 * p[1].dx, p[1].dy);
+    p[5] = new Offset(-1.0 * p[0].dx, p[0].dy);
+
+    // translate all points back to the absolute coordinate system.
+    for (int i = 0; i < p.length; i += 1)
+      p[i] += guest.center;
 
     return new Path()
-      ..lineTo(guest.center.dx - b, host.top)
+      ..lineTo(p[0].dx, p[0].dy)
+      ..quadraticBezierTo(p[1].dx, p[1].dy, p[2].dx, p[2].dy)
       ..arcToPoint(
-        new Offset(guest.center.dx + b, host.top),
+        p[3],
         radius: new Radius.circular(notchRadius),
         clockwise: false,
       )
+      ..quadraticBezierTo(p[4].dx, p[4].dy, p[5].dx, p[5].dy)
       ..lineTo(end.dx, end.dy);
   }
+
+  bool _notchAssertions(Rect host, Rect guest, Offset start, Offset end,
+    double fabRadius, double notchRadius) {
+    if (end.dy != host.top)
+      throw new FlutterError(
+        'The floating action button\'s notch maker must only be used for a notch in the top edge of the host.\n'
```

Comment:
The reference to "notch maker" is a little confusing here (and below). MIght be better to say: ... the y coordinate of the end of the FAB notch must match the top of the host.

Reply:
Changed to "The notch of the floating action button must end at the top edge of the host."